clippy_lints/src/methods/filter_map.rs

```diff
@@ -50,6 +52,197 @@ fn is_option_filter_map(cx: &LateContext<'_>, filter_arg: &hir::Expr<'_>, map_ar
+                    Some(higher::IfLetOrMatch::IfLet(sc, pat, then, Some(else_)))
+                        if let Some((ident, span)) = expr_uses_local(pat, then) => {
+                        (sc, else_, ident, span)
+                    },
+                    Some(higher::IfLetOrMatch::Match(sc, [arm1, arm2], MatchSource::Normal)) => {
+                        let is_variant_arm = |arm: &hir::Arm<'_>, other_arm: &hir::Arm<'_>| {
+                            // check if the other arm has pattern `_` and this arm references the local (modulo blocks)
+                            if let PatKind::Wild = other_arm.pat.kind {
+                                expr_uses_local(arm.pat, arm.body.peel_blocks())
+                            } else {
+                                None
+                            }
+                        };
+                        let (else_arm, (ident, span)) = is_variant_arm(arm1, arm2)
+                            .map(|data| (arm2, data))
+                            .or_else(|| is_variant_arm(arm2, arm1).map(|data| (arm1, data)))?;
+
+                        (sc, else_arm.body, ident, span)
+                    },
+                    _ => return None,
+                };
+
+                if path_to_local_id(scrutinee, map_param_id)
+                    // else branch should be a `panic!` or `unreachable!` macro call
+                    && let Some(mac) = root_macro_call(else_.peel_blocks().span)
+                    && (is_panic(cx, mac.def_id) || cx.tcx.opt_item_name(mac.def_id) == Some(sym::unreachable))
+                {
+                    Some(CheckResult::PatternMatching { variant_span, variant_ident })
+                } else {
+                    None
+                }
+            },
+        }
+    }
+
+    fn hir(cx: &LateContext<'tcx>, expr: &'tcx Expr<'tcx>, filter_param_id: HirId) -> Option<Self> {
+        if let ExprKind::MethodCall(path, receiver, [], _) = expr.kind
+            && let Some(recv_ty) = cx.typeck_results().expr_ty(receiver).peel_refs().ty_adt_def()
+            // TODO: we intentionally (?) do not check if `receiver` is a path expression, but maybe we should..?
+            // .filter(|x| effect(x).is_some()).map(|x| effect(x).unwrap()) to filter_map can have different behavior
+        {
+            if cx.tcx.is_diagnostic_item(sym::Option, recv_ty.did())
+                && path.ident.name == sym!(is_some)
+            {
+                Some(Self::IsSome { receiver })
+            } else if cx.tcx.is_diagnostic_item(sym::Result, recv_ty.did())
+                && path.ident.name == sym!(is_ok)
+            {
+                Some(Self::IsOk { receiver })
+            } else {
+                None
+            }
+        } else if let Some(macro_call) = root_macro_call(expr.span)
+            && is_matches_macro(cx, macro_call.def_id)
+            && let ExprKind::Match(scrutinee, arms, _) = expr.kind
```

Comment:
This can be

```suggestion
            && let ExprKind::Match(scrutinee, [arm, _], _) = expr.kind
```

The `_ => false` arm must come second